SalesforceCommerceCloud/pwa-kit
# [Circle CI] Add test generated project job on Windows

GUS: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000mBHOTYA4/view



# Description


The PR adds a new Circle CI job to generate a project from the monorepo on a Windows Circle CI machine.


### Problem
Previously the Verdaccio configuration was set to output the Verdaccio logs to the terminal. 
The `create-mobify-app-dev.js` script read the logs outputted to the terminal, waiting for Verdaccio to be up and running before proceeding with the rest of the promise chain. The script also created a `verdaccio.log` file with some of the logs printed in the terminal.

The script first publishes the existing versions  `pwa-kit-create-app` and `pwa-kit-react-sdk` monorepo packages to the local npm repository Verdaccio. Then generates a project from the packages that we just published to the local npm repository using the command `npx pwa-kit-create-app` and finally runs tests on the generated project.

A more detailed description of the `create-mobify-app-dev.js` script can be found [here](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/e130dd48d1b6f8c032e09051b8e719a8932cd218/packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js#L10).


That approach was not working on a Windows Circle CI machine. The process hung without any output for more than 10 min.
(Optional see failed test on previous commits of this branch).

Potentially the problem was related to the terminal output piping between the different processes run in the promise chain. The Verdaccio logs were overwritten and lost by the following processes output in the promises, ending in a situation where the script didn't output any output and nor the expected non-zero exit by Circle CI.


### Solution
The Verdaccio configuration has been updated to output all the Verdaccio's logs to a `verdaccio.log` file using the native Verdaccio functionality. 

The `create-mobify-app-dev.js` script now monitors the `verdaccio.log` file content and waits until Verdaccio is ready before continuing with the rest of the promise chain.

The generated project `npm install` progress is now visible like we did on V2 (BFH branch).

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes
- Added a new Circle CI job to generate a project from the monorepo on a Windows machine.
- Refactor a new `generate_and_test_project` step in the Circle CI config to be used by the test jobs (`generatedSFCCProjectTestWindows` and `generatedSFCCProjectTestWindows`) running on Linux and Windows machines.
- Changed Verdaccio configuration to output the logs to a `verdaccio.log` file.
- Updated Verdaccio to the latest v4 version.
- The `create-mobify-app-dev.js` script now monitors the `verdaccio.log` file instead of the terminal.
- Switched `shelljs` with Node `child_process` for some processed that were stuck on Windows.
- The generated project `npm install` progress is now visible.

# How to Test-Drive This PR

- Check the new Circle CI job `generatedSFCCProjectTestWindows` and verify that works as expected. https://circleci.com/gh/SalesforceCommerceCloud/pwa-kit/9978?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

(Optional) Manually run the `create-mobify-app-dev.js` script.

Mac:
```
GENERATOR_PRESET=test-project node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir generated-project
```

Windows:
```
set GENERATOR_PRESET=test-project
node packages\pwa-kit-create-app\scripts\create-mobify-app-dev.js --outputDir generated-project
```

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
